crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,1132 @@
+use std::fmt;
+
+use ruff_db::file_system::{FileSystemPath, FileSystemPathBuf};
+use ruff_db::vfs::{system_path_to_file, VfsFile};
+
+use crate::db::Db;
+use crate::module_name::ModuleName;
+use crate::typeshed::{LazyTypeshedVersions, TypeshedVersionsQueryResult};
+use crate::TargetVersion;
+
+/// Enumeration of the different kinds of search paths type checkers are expected to support.
+///
+/// N.B. Although we don't implement `Ord` for this enum, they are ordered in terms of the
+/// priority that we want to give these modules when resolving them,
+/// as per [the order given in the typing spec]
+///
+/// [the order given in the typing spec]: https://typing.readthedocs.io/en/latest/spec/distributing.html#import-resolution-ordering
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum ModuleResolutionPathBufInner {
+    Extra(FileSystemPathBuf),
+    FirstParty(FileSystemPathBuf),
+    StandardLibrary(FileSystemPathBuf),
+    SitePackages(FileSystemPathBuf),
+}
+
+impl ModuleResolutionPathBufInner {
+    fn push(&mut self, component: &str) {
+        if let Some(extension) = camino::Utf8Path::new(component).extension() {
+            match self {
+                Self::Extra(_) | Self::FirstParty(_) | Self::SitePackages(_) => assert!(
+                    matches!(extension, "pyi" | "py"),
+                    "Extension must be `py` or `pyi`; got `{extension}`"
+                ),
+                Self::StandardLibrary(_) => {
+                    assert!(
+                        matches!(component.matches('.').count(), 0 | 1),
+                        "Component can have at most one '.'; got {component}"
+                    );
+                    assert_eq!(
+                        extension, "pyi",
+                        "Extension must be `pyi`; got `{extension}`"
+                    );
+                }
+            };
+        }
+        let inner = match self {
+            Self::Extra(ref mut path) => path,
+            Self::FirstParty(ref mut path) => path,
+            Self::StandardLibrary(ref mut path) => path,
+            Self::SitePackages(ref mut path) => path,
+        };
+        inner.push(component);
+    }
+}
+
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub(crate) struct ModuleResolutionPathBuf(ModuleResolutionPathBufInner);
```

Comment:
I find it quite confusing in reading the code, and I think also it is error-prone, that we use the same type to represent both (a) a search path root, and (b) the path to an actual module. I think these are different things that need to obey different constraints (e.g. a search path root should never have any file extension at all) and provide different behaviors (e.g. a search path root shouldn't be mutable; we shouldn't ever push new components to it; many of the other methods provided here are also never used - and should never be used, because they aren't even meaningful - on a search root).

IMO it will be a lot clearer if we explicitly represent these as different types. Right now we only know which is which by context in the calling code; we get no help from the type system in differentiating two things that should never be confused with each other. I don't think it would even be that complex a change; it would just mean that in some cases where we currently "clone" a search root into an initial package path, we'd instead be explicitly constructing one type from the other.

Reply:
It might even make sense that a package/module path is always represented as simply a relative path and an Arc to a search root. Then the "kind" is already encoded in the search root kind, and we can also eliminate all these cases of needing to make sure the passed-in search root kind matches the module path kind.

Reply:
Alex's first version used different types for each path variant and it added a ton of boilerplate code (about 1000 lines) that made it very hard to find the "important" bits. I also think that it didn't add much value because the only places where the different path types were used were in the match arms of that given variant. In which case using different types doesn't protect from much. If you screw up the variant name, than it is as easy to screw up the path type as well (and it's a very easy error to spot). 

I would think differently if the paths were exposed externally, which they are not.

Reply:
I don't think "use different types for each path variant" is at all related to the suggestion I'm making here.

Reply:
Yeah, I agree that what you propose sounds cleaner, but I think what I have now is okay, and I'd like to land this now. If I have time next week, I'm happy to come back to this

Reply:
I've added a TODO